### ISSUE TITLE
backend/s3: `assume_role` and `assume_role_with_web_identity` ignores environment variables

### ISSUE THREAD

**Issue by jar-b** [hashicorp/terraform#33803]
### Terraform Version

```shell
v1.6.0-beta1
```

### Terraform Configuration Files

```terraform
terraform {
  backend "s3" {
    bucket = "jb-test"
    key    = "path/to/statefiles"
    region = "us-east-1"
  }
}
```

### Debug Output

https://github.com/hashicorp/terraform/pull/31276#issuecomment-1703530574

### Expected Behavior

The `assume_role` and `assume_role_with_web_identity` blocks should respect the corresponding AWS environment variables.

### Actual Behavior

Role assumption flow fails.

### Steps to Reproduce

1. Set `AWS_ROLE_ARN` environment variable (and `AWS_WEB_IDENTITY_TOKEN_FILE` if using web identity).
2. `terraform init`
3. Observe failure

### Additional Context

Originally reported in this comment: https://github.com/hashicorp/terraform/pull/31276#issuecomment-1703530574

### References

Relates #31276 
Relates #33730

**Comment by gdavison** [comment]
@manobi this should actually work, since the AWS SDK for Go handles the environment variables. I've created a branch that (re-)enables API request logging for the authentication flow at https://github.com/hashicorp/terraform/tree/s3/log-base.

Can you try again using that branch and with the environment variable `TF_LOG` set to `DEBUG` and share the log, please?

Can you also please share your backend configuration, please?

**Comment by manobi** [comment]
I would love to help with this test, but I only have this trust relationship for my CI/CD runners.

I can't try it until it's published as a docker image on docker hub or something like it.

**Comment by gdavison** [comment]
The logging update should be released in `v1.6.0-beta2`

**Comment by manobi** [comment]
Have it already being released to public docker registries?

**Comment by crw** [comment]
No, `v1.6.0-beta2` has not yet been built and released at all, thus it is not in Dockerhub. See also: https://hub.docker.com/r/hashicorp/terraform/tags

**Comment by gdavison** [comment]
Hi @manobi, `v1.6.0-beta3` is now on Dockerhub

**Comment by gdavison** [comment]
Hi @manobi. Now that v1.6 has been released, are you still seeing this problem? If so, can you please share your backend configuration and a debug log generated by setting the environment variable `TF_LOG` to `DEBUG`

**Comment by marshallford** [comment]
I'm seeing a similar issue with v1.6.5, however the steps to reproduce are a bit different (set `AWS_WEB_IDENTITY_TOKEN_FILE` env var, set [`role_arn`](https://developer.hashicorp.com/terraform/language/settings/backends/s3#role_arn-2) in backend config). 

```
Initializing the backend...
╷
│ Error: Missing Required Value
│ 
│   on main.tf line 9, in terraform:
│    9:   backend "s3" {
│ 
│ Exactly one of web_identity_token, web_identity_token_file must be set.
╵
```

Perhaps I'm barking up the wrong tree and this is a SDK issue, similar to: https://github.com/hashicorp/terraform-provider-aws/issues/27019.

**Comment by ekristen** [comment]
@crw @justinretzolk since you both responded to my issue yesterday in a related way, I wanted to bring this to your attention as soon as possible given that 1.10 now enforces the split of assume role and assume web identity, essentially assume with web identity is broken for 1.10 if you were previously using environment variables. 

```
Exactly one of web_identity_token, web_identity_token_file must be set.
```

I can confirm 100% that `AWS_WEB_IDENTITY_TOKEN_FILE` is set and this is an AWS SDK supported environment variable, so not only has this issue existed in Sept 2023, now that 1.10 forces use of the two assume role blocks, it's now broken hard in 1.10+.

**Comment by ekristen** [comment]
@crw can we get eyes on this, we do a lot of automated deployments and use automatic env vars to set roles and can't use this functionality due to this bug. I believe it's as simple as removing https://github.com/hashicorp/terraform/blob/main/internal/backend/remote-state/s3/backend.go#L522C1-L528C4 as the SDK should pick up the env var automatically if they aren't set.

**Comment by crw** [comment]
I'll re-raise it with the AWS Provider team here at HashiCorp.

**Comment by trc-ikeskin** [comment]
Any news on this issue?

**Comment by vegardx** [comment]
Can the documentation be updated to clarify this doesn't work? It would save people from wasting hours on configuration that can't succeed.

**Comment by Sharsie** [comment]
https://github.com/hashicorp/terraform-provider-aws/pull/48736 introduced an enhancement, released in v6.56.0, that now allows configuring the provider with `AWS_WEB_IDENTITY_TOKEN_FILE`. 

This makes it really difficult to use together with the backend config. Additionally, I'm mentioning this because not providing AWS_ROLE_ARN produces another error

<details open>
  <summary>Error without AWS_ROLE_ARN</summary>

```hcl
assume_role_with_web_identity = {
  role_arn = "arn:aws:iam::000000000000:role/my-role"
  web_identity_token_file = "/tmp/token"
  role_session_name = "_"
}
```

Executed with:

```console
export AWS_WEB_IDENTITY_TOKEN_FILE=/tmp/token
terraform init
...
| Error: role ARN is not set
```
</details>


<details open>
  <summary>Dropping hardcoded web_identity_token_file moves the error to already discussed validation error.</summary>

```hcl
assume_role_with_web_identity = {
  role_arn = "arn:aws:iam::000000000000:role/my-role"
  role_session_name = "_"
}
```

Executed with:

```console
export AWS_WEB_IDENTITY_TOKEN_FILE=/tmp/token
terraform init
...
| Exactly one of web_identity_token, web_identity_token_file must be set.
```
</details>

To work around this, the only way I found is to unset ENV vars, then set them again

```console
TMP_AWS_WEB_IDENTITY_TOKEN_FILE=$AWS_WEB_IDENTITY_TOKEN_FILE
TMP_AWS_ROLE_ARN=$AWS_ROLE_ARN
unset AWS_WEB_IDENTITY_TOKEN_FILE AWS_ROLE_ARN
terraform init
export AWS_WEB_IDENTITY_TOKEN_FILE=$TMP_AWS_WEB_IDENTITY_TOKEN_FILE
export AWS_ROLE_ARN=$AWS_ROLE_ARN
unset TMP_AWS_WEB_IDENTITY_TOKEN_FILE TMP_AWS_ROLE_ARN
```

**Comment by ekristen** [comment]
@crw you re-raised this over a year ago for me. This is still broken. Would really like to be able to upgrade past 1.9.
